File: 2nd.basic_drawing_examples/python/drawing.py
# USAGE
# python drawing.py

# Import the necessary packages
import numpy as np
import cv2

def draw_line():
	# Draw a green line from the top-left corner of our canvas
	# to the bottom-right
	green = (0, 255, 0)
	cv2.line(canvas, (0, 0), (300, 300), green)
	cv2.imshow("Canvas", canvas)
	cv2.waitKey(0)

def draw_rectangle():
	# Draw a green 50x50 pixel square, starting at 10x10 and
	# ending at 60x60
	green = (0, 255, 0)
	cv2.rectangle(canvas, (10, 10), (60, 60), green)
	cv2.imshow("Canvas", canvas)
	cv2.waitKey(0)

def draw_ellipse():
	red = (255, 0, 0)
	cx,cy = 201,113
	ax1,ax2 = 37,27
	angle = -108
	center = (cx,cy)
	axes = (ax1,ax2)
	cv2.ellipse(canvas, center, axes, angle, 0, 360, red, 2)

def draw_circle():
	# Reset our canvas and draw a white circle at the center
	# of the canvas with increasing radii - from 25 pixels to
	# 150 pixels
	(centerX, centerY) = (canvas.shape[1] / 2, canvas.shape[0] / 2)
	white = (255, 255, 255)
	for r in xrange(0, 175, 25):
		cv2.circle(canvas, (centerX, centerY), r, white)
	cv2.imshow("Canvas", canvas)
	cv2.waitKey(0)

def draw_fillpoly():
	pts = np.array([[10,5],[20,30],[70,20],[50,10]], np.int32)  
	pts = pts.reshape((-1,1,2))  
	cv2.polylines(canvas,[pts],True,(0,255,255),3)
	cv2.imshow("Canvas", canvas)
 	cv2.waitKey(0)
	
	

def draw_color():
	# Let's go crazy and draw 25 random circles
	for i in xrange(0, 25):
		# randomly generate a radius size between 5 and 200,
		# generate a random color, and then pick a random
		# point on our canvas where the circle will be drawn
		radius = np.random.randint(5, high = 200)
		color = np.random.randint(0, high = 256, size = (3,)).tolist()
		pt = np.random.randint(0, high = 300, size = (2,))
		# draw our random circle
		cv2.circle(canvas, tuple(pt), radius, color, -1)

	# Show our masterpiece
	cv2.imshow("Canvas", canvas)
	cv2.waitKey(0)

def put_text():
	font = cv2.FONT_HERSHEY_SIMPLEX  
	cv2.putText(canvas, 'Hello zhangyushan', (10,50), font, 1, (255,255,255) , 2, cv2.LINE_AA)
	#cv2.putText(canvas, "Hello Zhangyushan", (50,50), cv2.FONT_HERSHEY_SIMPLEX, .6, (0, 255, 0), 1, 2)
	cv2.imshow("Canvas", canvas)
	cv2.waitKey(0)


if __name__=="__main__":
	#ap = argparse.ArgumentParser()
 	#ap.add_argument("-i", "--image", required = True,
	#	help = "Path to the image")
	#args = vars(ap.parse_args())
	
	# Initialize our canvas as a 300x300 with 3 channels,
	# Red, Green, and Blue, with a black background
	canvas = np.zeros((400, 400, 3), dtype = "uint8")
	draw_line();
	draw_rectangle()
	draw_ellipse()
	draw_circle()
	draw_fillpoly()
	draw_color()
	put_text()

File: 2nd.basic_drawing_examples/cpp/drawing.cpp
#include <opencv2/core/core.hpp>
#include <opencv2/highgui/highgui.hpp>
#include <opencv2/line_descriptor/descriptor.hpp>
#include <unistd.h>
#include <iostream>
#include <stdio.h>  

using namespace cv;
using namespace std;


//draw a line
int draw_line(const	cv::Mat& image){
	line(image, Point( 15, 20 ), Point( 370, 50), Scalar( 110, 220, 0 ),  2, 8 );
 	imshow("Image", image);
	waitKey( 0 );
	return 0;
}

//draw a cicle
int draw_circle(Mat& image){
  	circle( image, Point(200, 200), 32.0, Scalar( 0, 0, 255 ), 1, 8 );
  	imshow("Image",image);
	waitKey(0);
	return 0;
}

//draw ellipse
int draw_ellipse(Mat& image){
	// Draw a ellipse 
	ellipse( image, Point( 200, 200 ), Size( 100.0, 160.0 ), 45, 0, 360, Scalar( 255, 0, 0 ), 1, 8 );
	ellipse( image, Point( 200, 200 ), Size( 100.0, 160.0 ), 135, 0, 360, Scalar( 255, 0, 0 ), 10, 8 );
	ellipse( image, Point( 200, 200 ), Size( 150.0, 50.0 ), 135, 0, 360, Scalar( 0, 255, 0 ), 1, 8 );
	imshow("Image",image);
	waitKey(0);
	return (0);
}


//draw rectangle
int draw_rectangle(Mat& image){
	// Draw a rectangle ( 5th argument is not -ve)
	rectangle( image, Point( 15, 20 ), Point( 70, 50), Scalar( 0, 55, 255 ), +1, 4 );
	imshow("Image1",image);
	// Draw a filled rectangle ( 5th argument is -ve)
	rectangle( image, Point( 115, 120 ), Point( 170, 150), Scalar( 100, 155, 25 ), -1, 8 );
	imshow("Image2",image);
	waitKey(0);
	return (0);
}

int draw_polygon(Mat& image){
 	/** Create some points */
	int w = 400;
	Point rook_points[1][20];
	rook_points[0][0] = Point( w/4.0, 7*w/8.0 );
	rook_points[0][1] = Point( 3*w/4.0, 7*w/8.0 );
 	rook_points[0][2] = Point( 3*w/4.0, 13*w/16.0 );
  	rook_points[0][3] = Point( 11*w/16.0, 13*w/16.0 );
  	rook_points[0][4] = Point( 19*w/32.0, 3*w/8.0 );
  	rook_points[0][5] = Point( 3*w/4.0, 3*w/8.0 );
  	rook_points[0][6] = Point( 3*w/4.0, w/8.0 );
  	rook_points[0][7] = Point( 26*w/40.0, w/8.0 );
  	rook_points[0][8] = Point( 26*w/40.0, w/4.0 );
  	rook_points[0][9] = Point( 22*w/40.0, w/4.0 );
  	rook_points[0][10] = Point( 22*w/40.0, w/8.0 );
  	rook_points[0][11] = Point( 18*w/40.0, w/8.0 );
  	rook_points[0][12] = Point( 18*w/40.0, w/4.0 );
  	rook_points[0][13] = Point( 14*w/40.0, w/4.0 );
  	rook_points[0][14] = Point( 14*w/40.0, w/8.0 );
  	rook_points[0][15] = Point( w/4.0, w/8.0 );
  	rook_points[0][16] = Point( w/4.0, 3*w/8.0 );
  	rook_points[0][17] = Point( 13*w/32.0, 3*w/8.0 );
  	rook_points[0][18] = Point( 5*w/16.0, 13*w/16.0 );
  	rook_points[0][19] = Point( w/4.0, 13*w/16.0) ;
 
  	const Point* ppt[1] = { rook_points[0] };
  	int npt[] = { 20 };
	fillPoly( image, ppt, npt, 1, Scalar( 255, 255, 255 ), 8 );
	imshow("Image",image);
	
	waitKey(0);
	return(0);
}

int put_text(Mat& image){
	putText(image, "Hello Zhangyushan", Point(50,100), FONT_HERSHEY_SIMPLEX, 1, Scalar(0,200,200), 4);
  imshow("Image",image);
 
  waitKey(0);
  return(0);
}

int main(int argc, char *argv[]) { 
	cout << endl;
    cout << "Cloning Module" << endl;
    cout << "---------------" << endl;
    cout << "Options: " << endl;
    cout << endl;
    cout << "1) Draw  Line " << endl;
    cout << "2) Draw Circle" << endl;
    cout << "3) Draw Ellipse " << endl;
    cout << "4) Draw Rectangle" << endl;
    cout << "5) Draw Polygon " << endl;
    cout << "6) Put Text " << endl;
	cout << "7) All Above" << endl;
    cout << endl;
    cout << "Press number 1-6 to choose from above techniques: ";
    int type = 1;
    cin >> type;
    cout << endl;

	Mat image = Mat::zeros( 400, 400, CV_8UC3 );
	switch(type) {
		case 1: 
			draw_line(image);
			break;
		case 2:
			draw_circle(image);
			break;
		case 3:
			draw_ellipse(image);
			break;
		case 4:
			draw_rectangle(image);
			break;
		case 5:
			draw_polygon(image);
			break;
		case 6 :
			put_text(image);
			break;
		case 7:
			draw_line(image);
			draw_circle(image);
			draw_ellipse(image);
			draw_rectangle(image);
			draw_polygon(image);
			put_text(image);
		default :
			break;
	}

  return(0);
}

